### DISCUSSION TITLE
Is there a way to keep track of an application that is playing audio in the background (Google Podcast)?

### DISCUSSION THREAD

**Discussion by ghost** [ActivityWatch/activitywatch#969] [Q&A]
I wanted to know if there is a way to keep track of application that is playing audio in the background, if there is a watcher for that or any default config i may be missing...

**Comment by BelKed** [comment]
As far as I know, it's not possible to track an application that's playing audio in the background.
ActivityWatch only tracks the focused window.

**Reply by ghost** [reply]
Yeah, i'm trying to search for a Watcher that is capable of doing so, or even trying to open an "issues ticket" asking for the feature to be built in, this is a quite helpful thing.

**Reply by 2e3s** [reply]
It can only be a new watcher. This is what would be used in Linux for instance: https://wiki.archlinux.org/title/MPRIS CLI example
```shell
dbus-send --print-reply --dest=org.mpris.MediaPlayer2.spotify /org/mpris/MediaPlayer2 org.freedesktop.DBus.Properties.Get string:"org.mpris.MediaPlayer2.Player" string:"Metadata"
```
Windows could do this with WinRT API, no idea about MacOS. Firefox and Chromium supports it as well.

**Reply by ghost** [reply]
What about keep track of any application running in the background by selecting it? For example, if I want to monitor Google Podcasts, I would add it to a list, and even if it isn't in focus, it would still be counted.

**Reply by 2e3s** [reply]
If these google podcasts report to a MPRIS implementation, i.e. if this command returns your player while playing podcasts, then it's much easier to track it with a separate watcher than a hacking a window watcher:
```
dbus-send --session --dest=org.freedesktop.DBus --type=method_call --print-reply /org/freedesktop/DBus org.freedesktop.DBus.ListNames | grep mpris
```

**Reply by ghost** [reply]
I mean this is a web app "PWA" for chromium, so it would return the web-browser info.

string "org.mpris.MediaPlayer2.chromium.instance13793"

**Reply by ghost** [reply]
I could try to run it under nativefier, but i don't think it is worth the extra gigs of ram and cpu usage.

**Reply by ghost** [reply]
The thing is spying on tasks running in the computer, for a specific selected window titles.

**Reply by 2e3s** [reply]
So you could run this command to figure out if the data about your podcast is available:
```
dbus-send --print-reply --dest=org.mpris.MediaPlayer2.chromium.instance13793 /org/mpris/MediaPlayer2 org.freedesktop.DBus.Properties.Get string:"org.mpris.MediaPlayer2.Player" string:"Metadata"
```
Replace `org.mpris.MediaPlayer2.chromium.instance13793` with whatever is returned though because this `instance13793` looks unique for each browser run. Usually, the metadata is available from browsers, like track title.

I will probably try to put up MPRIS watcher together in a few days because given my experience I feel like the implementation on top of DBus is faster than setting up a repository for such a watcher and only slightly longer than writing this comment.

**Reply by 2e3s** [reply]
@RyanOrigens You could try this https://github.com/2e3s/aw-watcher-media-player it's very raw, but I think it should report what you expect if you don't do something unusual. There is a binary attached to the release, it's built on ubuntu 22.04 docker.

**Reply by ghost** [reply]
Hello 2e3s,

Thank you for putting time into doing the watcher!

I've tried here and it kind of worked, the problem though is that keep track of the last played sound in the background, it doesn't separate the two instances. (Even if you have only one audio playing and you clicked a youtube video at last it would be counted as the youtube video), This is browser issue probably.

It would be fixed by running the web-app in a natifier separate app rather than a webapp inside your browser.

The second problem is it is capable of grabbing the current audio running name only.

This is the result

   variant       array [
         dict entry(
            string "mpris:length"
            variant                int64 1735079184
         )
         dict entry(
            string "mpris:trackid"
            variant                object path "/org/chromium/MediaPlayer2/TrackList/Track8BEE3B6DFC4D410CD11F0F03ED6A3BCE"
         )
         dict entry(
            string "xesam:album"
            variant                string ""
         )
         dict entry(
            string "xesam:artist"
            variant                array [
                  string "Grace to You on Oneplace.com"
               ]
         )
         dict entry(
            string "xesam:title"
            variant                string "“The Inspiration of Scripture” (#1)"
         )
      ]
      
      It misses the name of the window itself (Google Podcasts - Grace to You on Oneplace.com - “The Inspiration of Scripture” (#1))

**Reply by ghost** [reply]
The solution I'm searching for is a little more complex than that. I need to keep track of applications running in the background without reproducing any audio. Audio is just an example; in reality, I want to track how many hours I spend coding per day, and the best way to do that is by monitoring my usage of Vscode. The problem is, most of the time, I'm on Google searching for solutions, Stack Overflow, ChatGPT (ChatBot), or watching tutorials on YouTube. During all this time, Vscode might still be open, but I wouldn't be focused on it, so it wouldn't accurately count the time.

Of course, I could add these other activities to the list, but it would still be very inaccurate because I also use chatbots, YouTube, and Google for other things.

The solution I'm seeking is the ability to track specific applications running in the background by their title or class.

**Reply by 2e3s** [reply]
> The second problem is it is capable of grabbing the current audio running name only.
> It misses the name of the window itself

You could run it in Firefox, it gives a bit different title. But this is how it's supposed to work. There can hardly be a link between Media Player reporting and the program's active window's title. You might write a browser extension which would report to DBus like [Plasma Browser integration](https://community.kde.org/Plasma/Browser_Integration) (or use that extension too if you use KDE) or directly to ActivityWatch. But it's not very useful IMO.

> the problem though is that keep track of the last played sound in the background, it doesn't separate the two instances. (Even if you have only one audio playing and you clicked a youtube video at last it would be counted as the youtube video), This is browser issue probably.

This sounds like a bug or a potential code improvement, a major browser deserves a hardcoded hack, but I don't quite follow how to reproduce it. If you have time and interest, could you please open an [issue](https://github.com/2e3s/aw-watcher-media-player/issues) with the description? I would triage it from there. By the way, I've made a fix of another problem: https://github.com/2e3s/aw-watcher-media-player/releases/tag/v0.0.2

I had an idea to make this extension, so when I saw your comments, it was a sign to commit, thank you for the push.

> Of course, I could add these other activities to the list, but it would still be very inaccurate because I also use chatbots, YouTube, and Google for other things.

And you remember to close VS Code as soon as you start using chatbots, YouTube, and Google for other things, then open it when you use chatbots, YouTube, and Google for coding?

> I want to track how many hours I spend coding per day, and the best way to do that is by monitoring my usage of Vscode. The problem is, most of the time, I'm on Google searching for solutions, Stack Overflow, ChatGPT (ChatBot), or watching tutorials on YouTube. During all this time, Vscode might still be open, but I wouldn't be focused on it, so it wouldn't accurately count the time.

IMO you're overthinking it with an ad-hoc solution for a specific workflow at the time being. I used to overthink it as well and wasted some time on an unnecessary project. Similar situation to [this](https://www.explainxkcd.com/wiki/index.php/1172:_Workflow).
If opening and closing VS Code are good measures of your working time (and it's sitting open for that time), then you could just click a timer manually on and off when you open and close VS Code, or automate it with Shell script which runs with VS Code. ActivityWatch web UI has a limited timer, we could commit efforts into enabling it, I might make the timer for my little aw-qt replacement's tray as well.
Here's [my other answer](https://github.com/orgs/ActivityWatch/discussions/925#discussioncomment-7008169) on a related topic, there are other good and less technical, more organizational ways to GTD and track things down.

**Comment by iloveitaly** [comment]
Which platform? I've been thinking about how to do this for Mac OS.

**Reply by 2e3s** [reply]
There is a server for media players for Mac OS https://crates.io/crates/souvlaki which means that it's possible there as well. I might add Mac OS to https://github.com/2e3s/aw-watcher-media-player if I will be able to run Mac OS in a virtual machine.
